File: src/omnia/data/structure/observable/ObservableUndirectedGraph.java
package omnia.data.structure.observable;

import io.reactivex.rxjava3.core.Observable;
import omnia.data.structure.Set;
import omnia.data.structure.UndirectedGraph;

public interface ObservableUndirectedGraph<E> extends ObservableGraph<E>, UndirectedGraph<E> {

  @Override
  ObservableChannels<E> observe();

  interface ObservableChannels<E> extends ObservableGraph.ObservableChannels<E> {

    @Override
    Observable<? extends UndirectedGraph<E>> states();

    @Override
    Observable<? extends MutationEvent<E>> mutations();
  }

  interface MutationEvent<E> extends ObservableGraph.MutationEvent<E> {

    @Override
    UndirectedGraph<E> state();

    @Override
    Set<? extends ObservableGraph.GraphOperation<E>> operations();
  }
}

File: src/omnia/data/structure/observable/ObservableUndirectedGraph.kt
package omnia.data.structure.observable

import io.reactivex.rxjava3.core.Observable
import omnia.data.structure.Set
import omnia.data.structure.UndirectedGraph
import omnia.data.structure.observable.ObservableGraph.GraphOperation

interface ObservableUndirectedGraph<E : Any> : ObservableGraph<E>, UndirectedGraph<E> {

  override fun observe(): ObservableChannels<E>

  interface ObservableChannels<E : Any> : ObservableGraph.ObservableChannels<E> {

    override fun states(): Observable<out UndirectedGraph<E>>

    override fun mutations(): Observable<out MutationEvent<E>>
  }

  interface MutationEvent<E : Any> : ObservableGraph.MutationEvent<E> {

    override fun state(): UndirectedGraph<E>

    override fun operations(): Set<out GraphOperation<E>>
  }
}